### DISCUSSION TITLE
Question about usage of model.generate()

### DISCUSSION THREAD

**Discussion by ottarg** [pybamm-team/PyBaMM#4046] [Q&A] [answered]
Hi all, 

I am just getting familiar with PyBaMM. What I am trying to do is get a C or C++ version of a cell model, for use on an embedded system.

I see that PyBaMM models have methods for calling the CasADi code generation functionality. 
https://docs.pybamm.org/en/stable/_modules/pybamm/models/base_model.html#BaseModel.generate
https://web.casadi.org/docs/#generating-c-code

and it seems as though the intent of this CasADi functionality is indeed to support embedded systems, but it isn't immediately clear to me if I am understand this feature correctly.  Is the purpose of
`model.generate(filename, variable_names, input_parameter_order=None, cg_options=None)`
to be able to obtain C/C++ code that can run on a step-by-step basis? i.e. essentially to run in the same fashion as what is described in the BaseSolver.step() method? (https://docs.pybamm.org/en/stable/_modules/pybamm/solvers/base_solver.html#BaseSolver.step)
where I would then call this generated function each timestep with inputs being the previous system state and any further inputs like current and timestep? 

If so, is there any example showing this workflow? 

Thanks!

**Comment by valentinsulzer** [comment, marked as answer]
Once you have generated the code, you have to compile it yourself and then you can call functions from the generated file. See https://github.com/pybamm-team/PyBaMM/blob/44d946f42e17de45e0fa4d21e1fe40c45a3395c7/tests/unit/test_models/test_base_model.py#L813-L834

You are then responsible for writing your own solver and post-processing. There are no workflows for doing this in PyBaMM.

**Reply by ottarg** [reply]
Hi Valentin, thanks for the response!

This is very helpful, I was quite confused before about what the nature of this generated code was meant to be. (I.e., is it providing the system model + solver, only the system description, or maybe even something completely different like the full experiment/simulation as a hardcoded program with no option to change anything that would just run whichever load profile you had chosen etc.)

I was able to build that test case you linked as well as going through the exercise of generating code for the example model in this tutorial: https://docs.pybamm.org/en/stable/source/examples/notebooks/creating_models/2-a-pde-model.html
That is working for me and things are beginning to make sense. I still definitely have some self-study to do, but I think I can figure it out now.

**Reply by ottarg** [reply]
@valentinsulzer, as a follow-up question: In the test example you linked as well as with the other models I am trying to build, it seems as though the functions other than x0 and z0 _always_ want 3 inputs, aside from the list of input parameters? It's a bit confusing to me because in the test example you showed, its clear that the first 3 inputs are [t,a,b] and then the second array is [p,q].

But in other models, Like the one from the pde example, its as though the inputs are [t,[concentration states],?} and then the input parameters make sense. Is the first input always "t" no matter whether it is even used in the model or not? In the PDE example, t is not used and there are fewer than 3 "variables" defined, yet it continues to insist I give it 3 values as inputs to the generated functions.  In some testing, I tried introducing a third variable and just using it as a scalar somewhere in the model and when I do that, it does appear to be used from that third input. However, when I try to add a _fourth_ variable to the test script you linked, it doesn't allow me then to input 4 values in the call to the generated script. 

Am I missing something here? I am struggling to understand the pattern it is looking for on the function inputs

**Reply by valentinsulzer** [reply]
The inputs required are:
- t, a scalar
- x, the values of the ODE states, which can be a scalar or a vector. In the example I linked, it's a scalar (a)
- z, the values of the algebraic states, which can be a scalar or a vector. In the example I linked, it's a scalar (b)
- the inputs parameters, which can be a scalar or a vector. In the example I linked, it's a vector ([p, q])

For example, in a model with 3 ode states, 2 algebraic states, and 4 input parameters, this would be a valid function call:
`rhs_fun(0, [1,2,3], [4,5], [6,7,8,9])`

**Reply by ottarg** [reply]
Ah okay thank you! That makes sense

**Reply by ottarg** [reply]
Hello again @valentinsulzer, sorry to bother you further about this, but I have reached a somewhat confusing blocker and I am not sure whether I have done something wrong, or whether I am misunderstanding something.

So far, I have succeeded in generating code for the built in model 
`model = pybamm.lithium_ion.DFN()`
I obtain the code, compile it, and I can run the model in matlab using ode15s and everything seems to work quite well. However, when I try to run the model with a solver of my own, I have a strange issue. (important to note: I have very little experience with such things and may be making a foolish mistake)

I am trying to put together a simple Newton-Raphson implicit solver, which looks something like this:
```
for i = 2:numTimeSteps
    yPrev = y(i-1, :);
    % Newton-Raphson iteration to find yCurrent
    yGuess = yPrev; % Initial guess
    for j = 1:maxIterations% Could also use a convergence-based stopping criterion
        % Evaluate the combined system residual at the current guess
        F = daeResiduals(t(i), yGuess, yPrev, dt, p);
        
        % Calculate Jacobian and solve the linear system for updates
        J = jacobianFunction(t(i), yGuess, p);
        delta = -J \ F;
        yGuess = yGuess + delta;

        % Check convergence
        if norm(delta, 'inf') < 1e-6
            break;
        end
    end
    y(i, :) = yGuess;
end
```

Where the method `daeResiduals(t(i), yGuess, yPrev, dt, p)` uses the generated code for 

`codeGenDFN("rhs_",t,x,z,p)` and `codeGenDFN("alg_",t,x,z,p)` to construct the residuals, and the `jacobianFunction(t(i), yGuess, p)`
calculates the Jacobian using  `codeGenDFN("jac_rhs",t,x,z,p)` and  `codeGenDFN("jac_alg",t,x,z,p)`. 

#### My problem:
I am construction the Jacobian J as:
```
Jx = codeGenDFN("jac_rhs",t,x,z,p);
Jz = codeGenDFN("jac_alg",t,x,z,p);
J = [Jx;Jz];

```
The built-in DFN model has 860 ODE states, and 100 algebraic states and so the dimensions are: `Jx = [860x960]` and `Jz = [100x960]`

However, when I then execute the step `delta = -J \ F;`, I have the problem that my J is rank-deficient, and so the calculation goes crazy.  The ranks of the matrices are:
```
rank(full(Jx) = 859
rank(full(Jz) = 100
rank(full(J)) = 957
```
Now, what I am unsure about is: Is this expected? I am not familiar enough with the DFN model to know the answer.  What really confuses me though is that this seems to be very clearly the reason why my solution is not working, and yet when I use ode15s with the same Jacobian function, it works perfectly fine.  Is there something obvious I am missing here, like an influence of boundary conditions etc which resolve this?

**Reply by valentinsulzer** [reply]
I'm not totally surprised that the DFN is rank deficient. ode15s probably does some smart approximations or reformulations to work even with rank deficient matrices. I can't think of an easy solution unfortunately. You'll have to either figure out how to make DFN not be rank deficient, or write a DAE solver that works with rank deficient matrices. Maybe others in the pybamm community will have ideas, I'm not an expert in numerical algorithms

**Reply by ottarg** [reply]
Okay! Thanks for the help, I'm at least glad to know that there is not an obvious thing I am missing.

**Reply by syed0923** [reply]
Hi ottarg!

I am working on similar task such as converting the code into C /MATLAB compatible version.  If possible, can you share how you did that or what procedure you followed?

Thanks in advance!
